evm/src/fixed_recursive_verifier.rs

```diff
@@ -1071,14 +1080,18 @@ where
                     nonzero_pis.insert(block_hashes_keys.start + 8 * (i + 1) + j, targets[j]);
                 }
             }
+            let block_hashes_current_start = TrieRootsTarget::SIZE * 2
+                + BlockMetadataTarget::SIZE
+                + BlockHashesTarget::BLOCK_HASHES_SIZE
+                - 8;
             let cur_targets = h256_limbs::<F>(public_values.block_hashes.prev_hashes[255]);
             for i in 0..8 {
-                nonzero_pis.insert(block_hashes_keys.end + i, cur_targets[i]);
+                nonzero_pis.insert(block_hashes_current_start + i, cur_targets[i]);
             }
 
             // Initialize the block number.
-            let block_number_key = TrieRootsTarget::SIZE * 2 + 6;
-            nonzero_pis.insert(block_number_key, F::NEG_ONE);
+            let block_number_key = TrieRootsTarget::SIZE * 2 + 3;
```

Comment:
A comment explaining the `3` would be good.  And a test?

Reply:
I was discussing with the team of a way to refactor all those indices which are quite error prone. The 3 refers to the position to this field in `BlockMetadata` but tests not calling `prove_block` won't need it hence why it wasn't caught earlier.